metabase/metabase
# Display collection items metadata

It's really hard to know in Metabase right now who made or has been recently working with a given dashboard, question, or collection

This PR:

* adds the "Last edited at" and "Last edited by" columns to collection items lists
* allows sorting collection items by their type, name, and last edit time
* adds a note about the last edit to the saved question page header
* adds a note about the last edit to the dashboard page header
* closes #15773, closes #6750, closes #13435

Notes:

- "Last edited at" and "Last edited by" columns are not displayed for Pulses. We retrieve this info from the most recent revision, and Pulses don't have them
- pinned items' `collection_position` field won't affect their position in the list anymore
- items filtering by their type is replaced with sorting (questions first, dashboards first)
- backend PR: #16275 

## To Verify

1. Open a collection. The more different items it contains, the better
2. Ensure both pinned and not pinned tables are sorted alphabetically by name as a default
3. Click the "Name" column, the list has to be sorted in reversed alphabetical order
4. Click the "Type" column; the items have to be sorted in Dashboards-Pulses-Questions order
5. Click the "Type" column one more time; the items have to be sorted in Questions-Pulses-Dashboards order
6. Click the "Last edited at" column; the items have to be sorted by that column now
7. Try applying different sortings to pinned and not pinned tables. They have to be separately sortable and not affect one another

## How to Review

Had to make a few refactorings here and there, so we can switch from a list-based layout to a table-based. Sorry if it makes the review inconvenient.

It can be easier to review the PR by commits. I've tried to keep them as small as possible.

The overall approach was to create separate [BaseItemsTable](https://github.com/metabase/metabase/blob/feat/collections-metadata/frontend/src/metabase/collections/components/BaseItemsTable.jsx), [PinnedItemsTable](https://github.com/metabase/metabase/blob/feat/collections-metadata/frontend/src/metabase/collections/components/PinnedItemsTable.jsx) and [ItemsTable](https://github.com/metabase/metabase/blob/feat/collections-metadata/frontend/src/metabase/collections/components/ItemsTable.jsx) components replacing [ItemsList](https://github.com/metabase/metabase/blob/v0.39.3/frontend/src/metabase/collections/components/ItemList.jsx) and [PinnedItemsList](https://github.com/metabase/metabase/blob/v0.39.3/frontend/src/metabase/collections/components/PinnedItems.jsx), port lists' functionality commit-by-commit and delete the replaced components.

Also, listing the most noticeable changes below:

- split [EntityItem](https://github.com/metabase/metabase/blob/fb3b85f757ba8eeba87e1b114155e93ce6f96c65/frontend/src/metabase/components/EntityItem.jsx) into a few components (icon, name, action menu), so they can be rendered independently in different table cells. Commits: [#19c1c19](https://github.com/metabase/metabase/pull/16295/commits/19c1c19d8521fb8123926208d57ecbc7145fdfce) [#069a92b](https://github.com/metabase/metabase/pull/16295/commits/069a92b3c91f6824f029ffc4f69f73f59a017a78) [#b0e2c14](https://github.com/metabase/metabase/pull/16295/commits/b0e2c146cfa7ca594186d088619148abd2a7ec2c) [#b1c2587](https://github.com/metabase/metabase/pull/16295/commits/b1c2587e3390e624312b0ef87698181d4c2ef246) [#4bec79c](https://github.com/metabase/metabase/pull/16295/commits/4bec79cebf873d11ede75ac157b57c9e50e9b898)
- updated `ItemDragSource` used for drag-n-drop. `react-dnd` allows wrapping only native DOM elements with DND helpers, so `ItemDragSource` wrapped its children with a `div`. With the new layout, we now need to drag table rows (`td` elements). Rendering a `td` with `ItemDragSource` results in invalid HTML as table body's direct children should only include `td` elements. [This commit](https://github.com/metabase/metabase/commit/3cb1b90556ad53c2f8719cd09b5d0fc54dfde8ff) removes a wrapping `div` and [this commit](https://github.com/metabase/metabase/commit/a6f5689c077cad380e77c7d2bfec578c0f6bbf39) wraps another `ItemDragSource` child with a `div`
- updated [ItemDragLayer](https://github.com/metabase/metabase/blob/432c51d91df3ca8fb0509329de4ab6cd63205efc/frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx). Once a user starts to drag an item, the app renders its absolutely positioned copy that follows the user's cursor. As a single item's layout is updated, the dragged item view has to respect that. Had to basically render the whole HTML table without column heads, so  column widths are correct
- created [`collections/collection-items-listing.cy.spec.js` test file](https://github.com/metabase/metabase/blob/243c454172c786df184855eabfbc51c29b03245a/frontend/test/metabase/scenarios/collections/collection-items-listing.cy.spec.js) containing pagination and sorting test. So don't be surprised if you see pagination E2E tests are removed from `collections.cy.spec.js`, they're just moved :D 

## Demo

### Before

**Collection Items List**

<img width="1095" alt="List (before)" src="https://user-images.githubusercontent.com/17258145/120652475-7c9a1080-c488-11eb-8ada-254a20c7f049.png">

**Dashboard Header**

<img width="569" alt="Dashboar header (before)" src="https://user-images.githubusercontent.com/17258145/120652489-7efc6a80-c488-11eb-880f-9ff5d6c117f6.png">

**Question Header**

<img width="434" alt="Question header (before)" src="https://user-images.githubusercontent.com/17258145/120652493-7f950100-c488-11eb-86f9-5aeab0077dcc.png">


### After

**Collection Items Table**

https://user-images.githubusercontent.com/17258145/120652942-e74b4c00-c488-11eb-98f8-7d576381a3aa.mov

**Dashboard Header**

<img width="573" alt="Dashboar header (after)" src="https://user-images.githubusercontent.com/17258145/120652957-ea463c80-c488-11eb-8e87-8b65ce60d05b.png">

**Question Header**

<img width="598" alt="Question header (after)" src="https://user-images.githubusercontent.com/17258145/120652963-ec100000-c488-11eb-8590-5c361dbe60bf.png">
